api/src/test/java/org/pmiops/workbench/rdr/RdrExportServiceImplTest.java

```diff
@@ -0,0 +1,126 @@
+package org.pmiops.workbench.rdr;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyList;
+import static org.mockito.ArgumentMatchers.anyListOf;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.anyShort;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.sql.Timestamp;
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.pmiops.workbench.db.dao.RdrExportDao;
+import org.pmiops.workbench.db.dao.UserDao;
+import org.pmiops.workbench.db.dao.WorkspaceDao;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.model.Degree;
+import org.pmiops.workbench.rdr.api.RdrApi;
+import org.pmiops.workbench.test.FakeClock;
+import org.pmiops.workbench.workspaces.WorkspaceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Import;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+public class RdrExportServiceImplTest {
+  @Autowired private RdrExportService rdrExportService;
+
+  @MockBean private ApiClient mockApiClient;
+  @MockBean private RdrApi mockRdrApi;
+  @MockBean private RdrExportDao rdrExportDao;
+  @MockBean private UserDao mockUserDao;
+
+  private static final Instant NOW = Instant.now();
+  private static final Timestamp NOW_TIMESTAMP = Timestamp.from(NOW);
+  private static final FakeClock CLOCK = new FakeClock(NOW, ZoneId.systemDefault());
+
+  private DbUser dbUserWithEmail;
+  private DbUser dbUserWithoutEmail;
+
+  @TestConfiguration
+  @Import({RdrExportServiceImpl.class})
+  @MockBean({
+      RdrApi.class,
```

Comment:
Some of these (RdrApi, RdrExportDao, and UserDao) should be unnecessary since you use the inline @MockBean annotation above (which I tend to prefer, as it directly associates it with an instance variable if needed). In fact, I tend to use only the inline version, even for mock beans that aren't directly used in the test (since it avoids creating a situation where some mockbeans are defined in one place, and some in the other place).

I'd be happy to hear reasons for preferring a different approach, though... I'm definitely still honing my Spring test aesthetic.

Reply:
No reason - still don't have a good sense of how to make tests work in Spring. I'll remove from this declaration the beans I have a handle for - if I pull all of them out someone will delete them for being unused and then wonder why the test doesn't work...